compiler/src/iree/compiler/Codegen/Common/WorkgroupSpecializationPass.cpp

```diff
@@ -0,0 +1,255 @@
+//
+// Transformed output
+//   cond = (boundedTileSizeY != TileX) && (boundedTileSizeX != TileY) && ...
+//   scf.if cond
+//     distribution loops with static shapes with the tile size
+//   else
+//     distribution loops with dynamic shapes with the tile size
+//
+// Steps:
+//   1. bail out if the loop bounds are already a multiple of the tile size.
+//   2. create a condition for scf.if
+//   3. clone the body of the innermost loop in the else block.
+//   4. clone the body of the else block into the then block and update the
+//      bounded size ops with the constant tile size during the cloning.
+static void specializeDistributionLoops(
+    SmallVector<LoopTilingAndDistributionInfo> &infoVec) {
+  // get the scf for loop nests and their tile size.
+  SmallVector<scf::ForOp> distLoops;
+  SmallVector<int64_t> tileSizes;
+  SmallVector<AffineMinOp> minSizeOps;
+
+  // Distribution info vector has the innermost loop at index 0.
+  for (auto &info : infoVec) {
+    auto forOp = cast<scf::ForOp>(info.loop);
+    distLoops.push_back(forOp);
+    if (!info.tileSize) {
+      return;
+    }
+    int64_t tileSize = *info.tileSize;
+    tileSizes.push_back(tileSize);
+    SmallVector<AffineMinOp> boundedTileSizesOps =
+        getBoundedTileSizeOps(forOp, tileSize);
+    if (boundedTileSizesOps.empty()) {
+      // When there is no bounded tile size op, it means that the tensor
+      // size is already a multiple of tile size.
+      minSizeOps.push_back(AffineMinOp());
+    } else if (boundedTileSizesOps.size() != 1) {
+      // Supposed to see only one op for the bounded tile size if any.
+      return;
+    } else {
+      minSizeOps.push_back(boundedTileSizesOps[0]);
+    }
+  }
+
+  // Check the eligilbility. Unsupported cases are:
+  //   1. Dynamic cases
+  //   2. UB < Tile size
+  //   3. UB == a multiple of the tile size
+  bool isAlreadyMultiple = true;
+  for (unsigned i = 0, e = distLoops.size(); i != e; ++i) {
+    scf::ForOp loop = distLoops[i];
+    auto ubOp = loop.getUpperBound().getDefiningOp<arith::ConstantIndexOp>();
+    if (!ubOp) {
+      // TODO: handle dynamic cases by adding more code to check the
+      // conditions.
+      return;
+    }
+    int64_t ub = ubOp.value();
+    int64_t tileSize = tileSizes[i];
+    if (ub % tileSize != 0) {
+      isAlreadyMultiple = false;
+    }
+    if (ub < tileSize) {
+      return;
+    }
+  }
+
+  if (isAlreadyMultiple) return;
+
+  scf::ForOp innermostLoop = distLoops[0];
+  auto loc = innermostLoop.getLoc();
+  Block *block = innermostLoop.getBody();
+
+  OpBuilder builder(innermostLoop->getContext());
+  PatternRewriter::InsertionGuard guard(builder);
+
+  AffineMinOp minOp0 = minSizeOps[0];
+  if (minOp0) {
+    // Make sure the affine.min in the innermost loop is the first
+    // instruction in the body.
+    auto minOp = llvm::dyn_cast_or_null<AffineMinOp>(block->front());
+    if (!minOp || minOp != minOp0) {
+      minOp0->moveBefore(&block->front());
```

Comment:
Moving ops like this can cause SSA use-def violation. THis is working for now since the expressions used here are the induction variable of the loop and dim values that are outside the loop. The dim values could be generated inside the loop as well. Such IR movements can cause issues.

Also, I'd rather take the `builder` as an input. That in general avoids a lot of footguns. But in this particular case its OK.

Reply:
Updated to check if the minOp's operands are all defined outside the loop body. This should guarantee that the minOp can be moved to the first in the block. Thanks